crates/telio-traversal/src/upgrade_sync.rs

```diff
@@ -53,32 +75,49 @@ pub trait UpgradeSyncTrait {
     async fn request_upgrade(
         &self,
         public_key: &PublicKey,
-        remote_endpoint: SocketAddr,
-        local_direct_endpoint: (SocketAddr, Session),
-    ) -> Result<()>;
+        remote_endpoint: (SocketAddr, EndpointProvider),
+        local_direct_endpoint: (SocketAddr, EndpointProvider),
+        session: Session,
+    ) -> Result<bool>;
+
+    async fn downgrade(&self, public_key: &PublicKey) -> Result<()>;
 }
 
 pub struct UpgradeSync {
     task: Task<State>,
 }
 
+struct PendingSessionData {
+    local_ep: EndpointProvider,
+    remote_ep: EndpointProvider,
+}
+
 pub struct State {
     upgrade_request_publisher: chan::Tx<UpgradeRequestChangeEvent>,
     upgrade_intercoms: Chan<(PublicKey, UpgradeMsg)>,
-    upgrade_requests: HashMap<PublicKey, UpgradeRequest>,
+    upgrade_requests: HashMap<Direction<PublicKey>, UpgradeRequest>,
     expiration_period: Duration,
     upgrade_controller: Arc<dyn UpgradeController>,
     poll_timer: Interval,
     upgrade_decision_intercoms: Chan<(PublicKey, UpgradeDecisionMsg)>,
+    // Collection of not yet confirmed Upgrade requests
+    pending_direct_sessions: LruCache<Session, PendingSessionData>,
+    our_public_key: PublicKey,
+    connectivity_data_aggregator: Arc<ConnectivityDataAggregator>,
+    wireguard: Arc<dyn WireGuard>,
```

Comment:
Well, I'm not sure if we want to use wireguard interface here - if we need to create AnalyticsEvent here, maybe it would be better to change it to some simpler struct and gather this data in aggregator, where we already have this data?

Reply:
I'm not convinced. If I have to be honest, I think that its aggregator that should not know about wireguard, not UpgradeSync.

Reply:
It might feel this way, but aggregator has to have it anyway and here it is unnecessary when it already has it

Reply:
I don't think that the argument that Aggregator is already contaminated is an argument for putting in it even more wireguard. Since Aggregator is the first thing that depends on Wireguard interface in the telio-nurse, I think we should hope to remove that dependency in future.